docs/compilers/CSharp/Runtime Async Design.md

```diff
@@ -515,36 +602,152 @@ Translated C#:
-        [3] int32 _tmp3
+        [0] int32 _tmp1,
+        [1] int32 _tmp2,
+        [2] int32 _tmp3
     )
 
     IL_0000: ldc.i4.0
     IL_0001: newarr [System.Runtime]System.Int32
-    IL_0006: stloc.0
-    IL_0007: call int32 C::M2()
-    IL_000c: stloc.1
+    IL_0006: call int32 C::M2()
+    IL_000b: stloc.0
+    IL_000c: dup
     IL_000d: ldloc.0
-    IL_000e: ldloc.1
-    IL_000f: ldelem.i4
-    IL_0010: stloc.2
-    IL_0011: call modreq(class [System.Runtime]System.Threading.Tasks.Task<int32>) int32 C::M1()
-    IL_0016: stloc.3
-    IL_0017: ldloc.0
-    IL_0018: ldloc.1
-    IL_0019: ldloc.2
-    IL_001a: ldloc.3
-    IL_001b: add
-    IL_001c: stelem.i4
-    IL_001d: ret
+    IL_000e: ldelem.i4
+    IL_000f: stloc.1
+    IL_0010: call class [System.Runtime]System.Threading.Tasks.Task`1<int32> C::M1()
+    IL_0015: call !!0 [System.Runtime]System.Runtime.CompilerServices.RuntimeHelpers::Await<int32>(class [System.Runtime]System.Threading.Tasks.Task`1<!!0>)
+    IL_001a: stloc.2
+    IL_001b: ldloc.0
+    IL_001c: ldloc.1
+    IL_001d: ldloc.2
+    IL_001e: add
+    IL_001f: stelem.i4
+    IL_0020: ret
 }
 ```
+
+#### Await a non-Task/ValueTask
+
+For anything that isn't a `Task`, `Task<T>`, `ValueTask`, and `ValueTask<T>`, we instead use `System.Runtime.CompilerServices.RuntimeHelpers.AwaitAwaiterFromRuntimeAsync` or
+`System.Runtime.CompilerServices.RuntimeHelpers.UnsafeAwaitAwaiterFromRuntimeAsync`. These are covered below.
+
+##### Implementor of ICriticalNotifyCompletion
+
+`ICriticalNotifyCompletion` lowering is always preferred over `INotifyCompletion` lowering, when we statically know `ICriticalNotifyCompletion` is implemented by the expression.
+
+```cs
+var c = new C();
+await c;
+
+class C
+{
+    public class Awaiter : ICriticalNotifyCompletion
+    {
+        public void OnCompleted(Action continuation) { }
+        public void UnsafeOnCompleted(Action continuation) { }
+        public bool IsCompleted => true;
+        public void GetResult() { }
+    }
+
+    public Awaiter GetAwaiter() => new Awaiter();
+}
+```
+
+Translated C#:
+
+```cs
+var c = new C();
+_ = {
+    var awaiter = c.GetAwaiter();
+    if (!awaiter.IsCompleted)
+    {
+        System.Runtime.CompilerServices.RuntimeHelpers.UnsafeAwaitAwaiterFromRuntimeAsync<C.Awaiter>(awaiter);
+    }
+    awaiter.GetResult()
+};
+```
+
+```il
+{
+    .locals init (
+        [0] class C/Awaiter awaiter
+    )
+
+    IL_0000: newobj instance void C::.ctor()
+    IL_0005: callvirt instance class C/Awaiter C::GetAwaiter()
+    IL_000a: stloc.0
+    IL_000b: ldloc.0
+    IL_000c: callvirt instance bool C/Awaiter::get_IsCompleted()
+    IL_0011: brtrue.s IL_0019
+
+    IL_0013: ldloc.0
+    IL_0014: call void [System.Runtime]System.Runtime.CompilerServices.RuntimeHelpers::UnsafeAwaitAwaiterFromRuntimeAsync<class C/Awaiter>(!!0)
+
+    IL_0019: ldloc.0
+    IL_001a: callvirt instance void C/Awaiter::GetResult()
+    IL_001f: ret
+}
+```
+
+#### Implementor of INotifyCompletion
```

Comment:
>#### [](http://example.com/codeflow?start=0&length=4)

nit: should be one more indent to match previous section?

Reply:
```suggestion
##### Implementor of INotifyCompletion
```